src/coinsselectionalgorithm.cpp

```diff
@@ -0,0 +1,361 @@
+    delete[] tempSelection;
+    delete[] optimalSelection;
+}
+
+CAmount* CCoinsSelectionAlgorithm::PrepareAmounts(std::vector<std::pair<CAmount, size_t>> unsortedAmountsAndSizes)
+{
+    std::vector<std::pair<CAmount, size_t>> sortedAmountsAndSizes = unsortedAmountsAndSizes;
+    std::sort(sortedAmountsAndSizes.begin(), sortedAmountsAndSizes.end(), [](std::pair<CAmount, size_t> left, std::pair<CAmount, size_t> right) -> bool { return ( left.first > right.first); } );
+    CAmount* sortedAmounts = new CAmount[problemDimension];
+    for (int index = 0; index < problemDimension; ++index)
+    {
+        sortedAmounts[index] = sortedAmountsAndSizes[index].first;
+    }
+    return sortedAmounts;
+}
+
+size_t* CCoinsSelectionAlgorithm::PrepareSizes(std::vector<std::pair<CAmount, size_t>> unsortedAmountsAndSizes)
+{
+    std::vector<std::pair<CAmount, size_t>> sortedAmountsAndSizes = unsortedAmountsAndSizes;
+    std::sort(sortedAmountsAndSizes.begin(), sortedAmountsAndSizes.end(), [](std::pair<CAmount, size_t> left, std::pair<CAmount, size_t> right) -> bool { return ( left.first > right.first); } );
+    size_t* sortedSizes = new size_t[problemDimension];
+    for (int index = 0; index < problemDimension; ++index)
+    {
+        sortedSizes[index] = sortedAmountsAndSizes[index].second;
+    }
+    return sortedSizes;
+}
+
+void CCoinsSelectionAlgorithm::Reset()
+{  
+    for (int index = 0; index < problemDimension; ++index)
+    {
+        tempSelection[index] = false;
+        optimalSelection[index] = false;
+    }
+
+    optimalTotalAmount = 0;
+    optimalTotalSize = 0;
+    optimalTotalSelection= 0;
+
+    isRunning = false;
+    stopRequested = false;
+    solvingThread = nullptr;
+    hasCompleted = false;
+    #if COINS_SELECTION_ALGORITHM_PROFILING
+    executionMicroseconds = 0;
+    #endif
+}
+
+std::string CCoinsSelectionAlgorithm::ToString()
+{
+    return std::string("Input:")
+                                + "{targetAmount=" + std::to_string(targetAmount) +
+                                  ",targetAmountPlusOffset=" + std::to_string(targetAmountPlusOffset) +
+                                  ",availableTotalSize=" + std::to_string(availableTotalSize) + "}\n" +
+                       "Output:"
+                                + "{optimalTotalAmount=" + std::to_string(optimalTotalAmount) +
+                                  ",optimalTotalSize=" + std::to_string(optimalTotalSize) +
+                                  ",optimalTotalSelection=" + std::to_string(optimalTotalSelection) + "}\n";
+}
+
+void CCoinsSelectionAlgorithm::StartSolvingAsync()
+{
+    if (!isRunning)
+    {
+        if (solvingThread != nullptr)
+        {
+            // this can be the case of a new start on the same algorithm instance,
+            // hence a deletion of previous pointer is performed before reallocation
+            delete solvingThread;
+        }
+        solvingThread = new std::thread(&CCoinsSelectionAlgorithm::Solve, this);
+    }
+}
```

Comment:
Being `isRunning` not atomic and checked with `compareExchangeStrong` we could invoke this multiple times with the effect of deleting the pointer to active thread(s)

Reply:
Member `isRunning` qualified as atomic, moved `solvingThread` from raw pointer to smart pointer.